istio/src/main/scala/io/buoyant/k8s/istio/mixer/MixerApiRequests.scala

```diff
@@ -0,0 +1,121 @@
+import com.twitter.util.Duration
+import io.buoyant.k8s.istio._
+import istio.mixer.v1._
+
+import scala.io.Source
+
+object MixerApiRequests {
+  val requestIndexIsIrrelevant = Some(210283L)
+
+  private[this] lazy val istioGlobalDict: Seq[String] = {
+    val yaml = Source.fromInputStream(
+      getClass.getClassLoader.getResourceAsStream("mixer/v1/global_dictionary.yaml")
+    ).mkString
+
+    val mapper = new ObjectMapper(new YAMLFactory) with ScalaObjectMapper
+    mapper.registerModule(DefaultScalaModule)
+    mapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)
+
+    mapper.readValue[Seq[String]](yaml)
+  }
+
+  def mkReportRequest(
+    responseCode: ResponseCodeIstioAttribute,
+    requestPath: RequestPathIstioAttribute,
+    targetService: TargetServiceIstioAttribute,
+    sourceLabel: SourceLabelIstioAttribute,
+    targetLabel: TargetLabelsIstioAttribute,
+    duration: ResponseDurationIstioAttribute
+  ): ReportRequest = {
+
+    val dictionary = mkDictionary(requestPath, targetService, sourceLabel, targetLabel)
+
+    val updated = mkAttributesUpdate(dictionary, Seq(responseCode, requestPath, targetService, sourceLabel, targetLabel, duration))
+    ReportRequest(requestIndexIsIrrelevant, Some(updated))
+  }
+
+  def mkCheckRequest(istioREquest: IstioRequest[_]) = {
+    val dictionary = mkDictionary(istioREquest.requestedPath, istioREquest.targetService, istioREquest.sourceLabel, istioREquest.targetLabel)
+    val requestAttributes = Seq(istioREquest.requestedPath, istioREquest.sourceLabel, istioREquest.targetLabel, istioREquest.targetService)
+    val attributesUpdate = mkAttributesUpdate(dictionary, requestAttributes)
+    CheckRequest(requestIndexIsIrrelevant, Some(attributesUpdate))
+  }
+
+  private def mkDictionary(
+    requestPath: RequestPathIstioAttribute,
+    targetService: TargetServiceIstioAttribute,
+    sourceLabel: SourceLabelIstioAttribute,
+    targetLabel: TargetLabelsIstioAttribute
+  ) = {
+    val targetLabelApp = targetLabel.value.getOrElse("app", "")
+    val targetLabelVersion = targetLabel.value.getOrElse("version", "")
+    val sourceLabelApp = sourceLabel.value.getOrElse("app", "")
+    val sourceLabelVersion = sourceLabel.value.getOrElse("version", "")
+
+    //TODO: when we first added Istio integration, this was required to make the integration work. We need to double-check if we can avoid sending these extra entries
+    val customWords = Seq(
+      "app",
+      "version",
+      requestPath.name,
+      targetService.name,
+      targetLabelApp,
+      targetLabelVersion,
+      sourceLabelApp,
+      sourceLabelVersion
+    ) ++ sourceLabel.value.keySet ++ targetLabel.value.keySet
+    customWords
+  }
+
+  private def mkAttributesUpdate(customAttributeNames: Seq[String], attributes: Seq[IstioAttribute[_]]) = {
+
+    // TODO: eventually we should not have to send istioGlobalDict over the wire,
+    // and instead use positive index integers
```

Comment:
I'm ignorant about Scala: when you retrieve indices in dictionary with `indexOf`, where're you flipping the index around to be negative? You want them to be negative since you're indexing into the local dictionary instead of the global one.

Reply:
Thanks so much, Zack!

1. Uhm.. the comment was about `mixerclient`, I probably misread something on the code. This iteration doesn't have a cache yet, but I will make sure to follow the same logic when we add it 👍 
2. This is brilliant, thank you. As I understand it, our first integration code was tested against an early alpha release and as everything was moving so fast we decided to send a lot of stuff just to make sure we didn't miss on anything that would become important or mandatory. I'll change it to the list you provided!
3. Our current integration runs against an older [schema that didn't have the negative index logic](https://github.com/istio/api/blob/ae97de1806d1802b65b6a0846f5498b7575f669e/mixer/v1/attributes.proto). I'll update the schemas as we move our integration to 0.2.x